tracing/src/main/java/com/palantir/tracing/SpanToken.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tracing;
+
+import java.io.Closeable;
+
+/**
+ * Closeable marker around a tracing span operation. This object should be used in a try/with block.
+ */
+public interface SpanToken extends Closeable {
```

Comment:
SpanToken vs TraceToken? Closing this represents closing a span, however it also cleans up the entire trace state. Thoughts on a better name? This class is unnecessary if we use the `Runnable`/`Callable` approach.

Reply:
I think Span is most correct here, these may be nested spans which are part of the same trace.

Reply:
why is a `SpanToken` required? (why can't we use `OpenSpan`?)

Reply:
OpenSpan is a data object, I want to expose an object whose only job is to complete an operation, with the ability to add validation that we have done so correctly.
We could use the static `Tracer.completeSpan()` method instead for this case because we do need to clean up thread state, however I've fixed too many tracing thread state leaks to be comfortable with that approach. Using the proposed API we can validate that we are closing the same span that we opened, which is not something we can do using the existing API.

Reply:
Why do we even need `SpanToken`? You can just return a `Closeable` anywhere you return `SpanToken`?

Reply:
Closeable can throw IOException